zcm/julia/ZCM.jl

```diff
@@ -0,0 +1,334 @@
+    data      ::Ptr{UInt8};
+end
+
+end
+
+# RRR: make variable naming consistent
+# RRR: improve orgainzation and make header comments more visible
+# Exported Objects and Methods
+type RecvBuf
+    rbuf  ::Ptr{Native.RecvBuf};
+    utime ::Int64;
+    data  ::Array{UInt8};
+
+    function RecvBuf(rbuf::Ptr{Native.RecvBuf})
+        instance  = new();
+        instance.rbuf = rbuf;
+
+        if (rbuf != C_NULL)
+            loadedRbuf = unsafe_load(rbuf);
+
+            instance.utime = loadedRbuf.recv_utime;
+
+            if (loadedRbuf.data != C_NULL)
+                instance.data = unsafe_wrap(Array, loadedRbuf.data, loadedRbuf.data_size);
+            else
+                instance.data = [];
+            end
+        end
+
+        # destruction of the underlying Native.RecvBuf pointer is handled by the C lib internally
+
+        return instance;
+    end
+end
+export RecvBuf;
+
+type __zcm_handler
+    handler::Any;
+    usr::Any;
+
+    # handler must have the function or functor prototype :
+    #   (rbuf::RecvBuf, channel::String, msg::MsgType, usr)
+    __zcm_handler(handler, usr) = new(handler, usr);
+end
+
+function __zcm_handler_wrapper(rbuf::Ptr{Native.RecvBuf}, channel::Cstring, usr_::Ptr{Void})
+    usr = unsafe_pointer_to_objref(usr_)::__zcm_handler;
+    usr.handler(RecvBuf(rbuf), unsafe_string(channel), usr.usr);
+    return nothing;
+end
+
+type Sub
+    sub::Ptr{Native.Sub};
+
+    # References to protect our callbacks from Julia's garbage collector
+    handler_jl::__zcm_handler;
+    handler_c ::Ptr{Void};
+    uv_handler::Ptr{Void};
+    uv_wrapper::Ptr{Native.UvSub};
+
+    Sub(sub::Ptr{Native.Sub}, handler_jl::__zcm_handler, handler_c::Ptr{Void},
+        uv_handler::Ptr{Void}, uv_wrapper::Ptr{Native.UvSub}) =
+        new(sub, handler_jl, handler_c, uv_handler, uv_wrapper);
+end
+
+type Zcm
+    zcm::Ptr{Native.Zcm};
+
+    # http://docs.julialang.org/en/stable/manual/calling-c-and-fortran-code/
+    # http://julialang.org/blog/2013/05/callback
+
+    # RRR: make prints throughout this file optional. Note that printing in a finalizer is not
+    #      ok because it can cause errors on exit if stdout has already been cleaned up
+    function Zcm(url::AbstractString)
+        instance = new();
+        instance.zcm = ccall(("zcm_create", "libzcm"), Ptr{Native.Zcm}, (Cstring,), url);
+
+        # user can force cleanup of their instance by calling `finalize(zcm)`
+        finalizer(instance, function(zcm::Zcm)
+                                if (zcm.zcm != C_NULL)
+                                    ccall(("zcm_destroy", "libzcm"), Void,
+                                          (Ptr{Native.Zcm},), zcm.zcm);
+                                    zcm.zcm = C_NULL;
+                                end
+                            end);
+
+        return instance;
+    end
+
+    function Zcm() return Zcm(""); end
+end
+export Zcm;
+
+function encode end
+function decode end
+
+function good(zcm::Zcm)
```

Comment:
`isgood` would be more standard Julia naming, but I understand if you want to make things as close as possible to the other languages.